### PR TITLE
Proposal for deterministic program generated Pubkey's that can be used only by programs to create signatures in process_instruction.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Programs cannot receive assets, or take "authority" without the ability to generate signatures.

#### Summary of Changes

Propose a mechanism for programs to generate addresses, and sign for those addresses when using cross program instruction invocations.

Fixes #

tag: @garious @jackcmay

### CODE REVIEW

File: book/src/proposals/program-keys-and-signatures.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+program specific addresses and programmatically set
+`KeyedAccount.is_signer` for those addresses when `process_instruction()`
+is invoked.
+
+### Deterministic Pubkey Addresses for Programs
+
+Deterministic pubkey addresses for programs follow a similar derivation
+path as Accounts created with `SystemInstruction::CreateAccountWithSeed`.
+
+```text
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: Use `rust,ignore` to syntax highlight

File: book/src/proposals/program-keys-and-signatures.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+use system_instruction::create_address_with_seed;
+
+pub fn derive_program_address(
+    program_id: &Pubkey,
+    keyword, &str,
+) -> Result<Pubkey, SystemError> {
+
+    //generate a deterministic base for all program addresses
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: Put a space after `//` to disambiguate from commented out code.  Capitalize the first word.

File: book/src/proposals/program-keys-and-signatures.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+## Proposed Solution
+
+The goal of this design is to allow programs to generate deterministic
+program specific addresses and programmatically set
+`KeyedAccount.is_signer` for those addresses when `process_instruction()`
+is invoked.
+
+### Deterministic Pubkey Addresses for Programs
+
+Deterministic pubkey addresses for programs follow a similar derivation
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: This is where you should introduce `CreateAddressWithSeed` and highlight that subtle difference, the word `Address`, not `Account` and that unlike everything else in that module, returns a `Pubkey` and not an `Instruction`.

File: book/src/proposals/program-keys-and-signatures.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+pub fn derive_program_address(
+    program_id: &Pubkey,
+    keyword, &str,
+) -> Result<Pubkey, SystemError> {
+
+    //generate a deterministic base for all program addresses
+    let base = create_address_with_seed(program_id,
+            &"derived program base address" , program_id)?;
+
+    //generate the keyword specific address
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: Hyphenate "keyword-specific"

File: book/src/proposals/program-keys-and-signatures.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    create_address_with_seed(&base, keyword, program_id);
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: That semicolon shouldn't be there.  And needs to be wrapped in an `Ok()`?

File: book/src/proposals/program-keys-and-signatures.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+address as if it had a private key to sign the transaction.
+
+```text
+fn transfer_one_token_from_escrow(
+    program_id: &Pubkey,
+    keyed_accounts: &[KeyedAccount]
+) -> Result<()> {
+
+    //deterministically derive the escrow key
+    let escrow = derive_program_address(program_id, &"escrow");
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: `escrow` -> `escrow_pubkey`

File: book/src/proposals/program-keys-and-signatures.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    //user supplies the destination
+    let alice_pubkey = keyed_accounts[1].key;
+
+    //create the transfer instruction
+    let instruction = token_instruction::transfer(&escrow, &alice, 1);
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: `alice` is undefined

File: book/src/proposals/program-keys-and-signatures.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    //Sign it with the key keyword.
+    //The runtime deterministically derives the key from the current
+    //program id and the supplied keyword.
+    //If the derived key matches a key in the instruction
+    //the `is_signed` flag is set.
+    sign_instruction(&instruction, &"escrow")?;
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: Need to define the function signature.  Unclear what the "escrow" string is.  Unclear why this should work - where's the private key?

File: book/src/proposals/program-keys-and-signatures.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    //deterministically derive the escrow key
+    let escrow = derive_program_address(program_id, &"escrow");
+
+    //user supplies the destination
+    let alice_pubkey = keyed_accounts[1].key;
+
+    //create the transfer instruction
+    let instruction = token_instruction::transfer(&escrow, &alice, 1);
+
+    //Sign it with the key keyword.
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: No previous mention of a "key" keyword, nor is it used below.

File: book/src/proposals/program-keys-and-signatures.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    //The runtime deterministically derives the key from the current
+    //program id and the supplied keyword.
+    //If the derived key matches a key in the instruction
+    //the `is_signed` flag is set.
+    sign_instruction(&instruction, &"escrow")?;
+
+    process_instruction(&instruction)?;
```
Comment (garious) [inline book/src/proposals/program-keys-and-signatures.md]: That semicolon shouldn't be there.

File: book/src/proposals/program-keys-and-signatures.md (hunk 3 above)
Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: hmm, what do you imagine is CreateAddressWithSeed, a new system instruction?  That's not what's being proposed here.

Comment (garious) [reply book/src/proposals/program-keys-and-signatures.md]: You wrote `use system_instruction::create_address_with_seed`.

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: That’s a utility function defined in that module for generating the address that’s used in CreateAccountWithSeed

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: Your confusion is inspiring :). There is an alternative design that i think will be much cleaner.

**Comment by aeyakovenko** [comment]
@jackcmay @garious bump, would love to have @jackcmay  grok this as he is working on process_instruction

File: book/src/proposals/program-keys-and-signatures.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+a destination address.
+
+```text
+//deterministically derive the escrow key
+let escrow_pubkey = derive_program_address(&escrow_program_id, &"escrow");
+let message = Message::new(vec![
+    token_instruction::transfer(&alice_pubkey, &escrow_pubkey, 1),
+]);
+//transfer 1 token to escrow
+client.send_message(&[&alice_pubkey, &escrow_pubkey], &message);
```
Comment (jackcmay) [inline book/src/proposals/program-keys-and-signatures.md]: @aeyakovenko Shouldn't this be:
`client.send_message(&[&alice_pubkey], &message);`
The client can't sign as the escrow correct?

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: @jackcmay they need the escrow pubkey in the TX

Comment (jackcmay) [reply book/src/proposals/program-keys-and-signatures.md]: @aeyakovenko 

send_message:
`fn send_message<T: Signers>(&self, keypairs: &T, message: Message) -> Result<Signature>;`

Aka, takes signer keypair and the message.  The escrow pubkey is communicated in the message/instruction

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: Ah, my bad.

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: fixed

Comment (aeyakovenko) [reply book/src/proposals/program-keys-and-signatures.md]: @jackcmay might be better if each key was defined by base + keyword or a slice of keywords.  I imagine the programs are going to have to use the users key as an identifier along with some symbolic name.  So something like <user pubkey>/<stake authority>,<withdraw authority>

**Comment by jackcmay** [comment]
@aeyakovenko Wana merge this in?  I can update it after

**Pull request merged**